## Домашнее задание к лекции "Статистика. Практика"
### Задание 1
Вернемся к набору данных о видеоиграх.

Ответьте на следующие вопросы:

1) Как критики относятся к спортивным играм?
2) Критикам нравятся больше игры на PC или на PS4?
3) Критикам больше нравятся стрелялки или стратегии?

Для каждого вопроса:

сформулируйте нулевую и альтернативную гипотезы;
выберите пороговый уровень статистической значимости;
опишите полученные результаты статистического теста.

In [1]:
import pandas as pd
import scipy.stats as stats

In [2]:
data1 = pd.read_csv('video_games_sales.txt')

In [3]:
data1.head()

Rank                           Name                      basename  \
0     1                     Wii Sports                    wii-sports   
1     2              Super Mario Bros.              super-mario-bros   
2     3                 Mario Kart Wii                mario-kart-wii   
3     4  PlayerUnknown's Battlegrounds  playerunknowns-battlegrounds   
4     5              Wii Sports Resort             wii-sports-resort   

      Genre ESRB_Rating Platform         Publisher         Developer  \
0    Sports           E      Wii          Nintendo      Nintendo EAD   
1  Platform         NaN      NES          Nintendo      Nintendo EAD   
2    Racing           E      Wii          Nintendo      Nintendo EAD   
3   Shooter         NaN       PC  PUBG Corporation  PUBG Corporation   
4    Sports           E      Wii          Nintendo      Nintendo EAD   

   VGChartz_Score  Critic_Score  ...  NA_Sales  PAL_Sales  JP_Sales  \
0             NaN           7.7  ...       NaN        NaN       NaN   
1             NaN          10.0  ...       NaN        NaN       NaN   
2             NaN           8.2  ...       NaN        NaN       NaN   
3             NaN           NaN  ...       NaN        NaN       NaN   
4             NaN           8.0  ...       NaN        NaN       NaN   

   Other_Sales    Year  Last_Update  \
0          NaN  2006.0          NaN   
1          NaN  1985.0          NaN   
2          NaN  2008.0  11th Apr 18   
3          NaN  2017.0  13th Nov 18   
4          NaN  2009.0          NaN   

                                                 url  status Vgchartzscore  \
0  http://www.vgchartz.com/game/2667/wii-sports/?...       1           NaN   
1  http://www.vgchartz.com/game/6455/super-mario-...       1           NaN   
2  http://www.vgchartz.com/game/6968/mario-kart-w...       1           8.7   
3  http://www.vgchartz.com/game/215988/playerunkn...       1           NaN   
4  http://www.vgchartz.com/game/24656/wii-sports-...       1           8.8   

                                         img_url  
0  /games/boxart/full_2258645AmericaFrontccc.jpg  
1                   /games/boxart/8972270ccc.jpg  
2  /games/boxart/full_8932480AmericaFrontccc.jpg  
3  /games/boxart/full_8052843AmericaFrontccc.jpg  
4  /games/boxart/full_7295041AmericaFrontccc.jpg  

[5 rows x 23 columns]

# Как критики относятся к спортивным играм?

1) Нужно сформулировать нулевую и альтернативную гипотезы;
H0: средняя оценка критиками спортивных игр не отличается от оценок других жанров
H1: средняя оценка критиками спортивных игр различается от оценок других жанров
2) Определить пороговый уровень статистической значимости;
возьмем стандартный уровень в 0.05
3) Сколько у нас выборок (какой вариант теста используем)?
используем двухвыборочный тест
3) Нужен ли односторонний, либо двухсторонний вариант теста?
используем двухсторонний вариант теста (и негативная и позитивная оценка)
5) Определить, равна ли дисперсия групп.
используем критерий Уэлча

In [4]:
# Создадим датасет только с оценками по спортивным играм
data_sport = data1[['Critic_Score']].loc[data1.Genre == 'Sports'].dropna()
# Создадим датасет только с оценками по играм остальных жанров
data_no_sport = data1[['Critic_Score']].loc[data1.Genre != 'Sports'].dropna()

Примечание: решила удалить строки без оценки и не заменять их на средние, например по жанру, так как на мой взгляд, это негативно повлияет на результаты сравнения. 

In [5]:
#Задаем уровень значимости
alpha = 0.05
#Применяем двухсторонний двухвыборочный критерий Стьюдента
result = stats.ttest_ind(data_sport['Critic_Score'], data_no_sport['Critic_Score'], equal_var=False)
print(result)
if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, средняя оценка критиками спортивных игр различается от оценок других жанров')
else:
    print('Не отвергаем нулевую гипотезу')

Ttest_indResult(statistic=1.467182255281725, pvalue=0.1426953974970837)
Не отвергаем нулевую гипотезу


Вывод: p_value > 0.05 -> не отвергаем нулевую гипотезу о равенстве средних оценок спортивных и остальных игр, 
    при этом средняя оценка по спортивным играм немного выше средней оценки игр других жанров.

# 2. Критикам нравятся больше игры на PC или на PS4?
1) Нужно сформулировать нулевую и альтернативную гипотезы;
H0: средняя оценка критиками игр на PC и PS4 одинакова
H1: средняя оценка критиками игр на PC и PS4 различается
2) Определить пороговый уровень статистической значимости;
возьмем стандартный уровень в 0.05
3) Сколько у нас выборок (какой вариант теста используем)?
используем двухвыборочный тест
3) Нужен ли односторонний, либо двухсторонний вариант теста?
используем двухсторонний вариант теста (так как нас интересует игры на какой платформе больше нравится критикам)
5) Определить, равна ли дисперсия групп.
используем критерий Уэлча

In [6]:
# Создадим датасет только с оценками критиков по играм на PC
data_PC = data1[['Critic_Score']].loc[data1.Platform == 'PC'].dropna()
# Создадим датасет только с оценками критиков по играм на PS4
data_PS4 = data1[['Critic_Score']].loc[data1.Platform == 'PS4'].dropna()

Примечание: решила удалить строки без оценки и не заменять их на средние, например по жанру, так как на мой взгляд, это негативно повлияет на результаты сравнения.

In [7]:
#Задаем уровень значимости
alpha = 0.05
#Применяем односторонний двухвыборочный критерий Стьюдента
result = stats.ttest_ind(data_PC['Critic_Score'], data_PS4['Critic_Score'], equal_var=False)
print(result)
if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, средняя оценка критиками игр на PC различается от оценок игр на PS4')
else:
    print('Не отвергаем нулевую гипотезу')

Ttest_indResult(statistic=-2.7394476056351627, pvalue=0.006931808250254211)
Отвергаем нулевую гипотезу, средняя оценка критиками игр на PC различается от оценок игр на PS4


Вывод: p_value < 0.05 -> отвергаем нулевую гипотезу о равенстве средних оценок критиков игр на PC и PS4, при этом средняя оценка на PC ниже средней оценки игр на PS4. С высокой долей статзначимости можем утверждать, что критикам больше нравятся игры на PS4.

# 3) Критикам больше нравятся стрелялки или стратегии?
1) Нужно сформулировать нулевую и альтернативную гипотезы; H0: средняя оценка критиками "стрелялок" и "стратегий" одинакова H1: средняя оценка критиками "стрелялок" и "стратегий" различается 2) Определить пороговый уровень статистической значимости; возьмем стандартный уровень в 0.05 3) Сколько у нас выборок (какой вариант теста используем)? используем двухвыборочный тест 3) Нужен ли односторонний, либо двухсторонний вариант теста? используем двухсторонний вариант теста (так как нас интересует какой вид игры больше нравится критикам) 5) Определить, равна ли дисперсия групп. используем критерий Уэлча

In [8]:
# Создадим датасет только с оценками критиков по играм на PC
data_Shooter = data1[['Critic_Score']].loc[data1.Genre == 'Shooter'].dropna()
# Создадим датасет только с оценками критиков по играм на PS4
data_Strategy = data1[['Critic_Score']].loc[data1.Genre == 'Strategy'].dropna()

Примечание: решила удалить строки без оценки и не заменять их на средние, например по жанру, так как на мой взгляд, это негативно повлияет на результаты сравнения.

In [9]:
#Задаем уровень значимости
alpha = 0.05
#Применяем односторонний двухвыборочный критерий Стьюдента
result = stats.ttest_ind(data_Shooter['Critic_Score'], data_Strategy['Critic_Score'], equal_var=False)
print(result)
if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, средняя оценка критиками "стрелялок" различается от оценок "стратегий"')
else:
    print('Не отвергаем нулевую гипотезу')

Ttest_indResult(statistic=-1.6073949711166526, pvalue=0.10838786414223071)
Не отвергаем нулевую гипотезу


Вывод: p_value > 0.05 -> не отвергаем нулевую гипотезу о равенстве средних оценок критиков игр на PC и PS4.
С недостаточным уровнем статзначимости можем сказать, что средняя оценка стратегий выше средней оценки стрелялок.

### Задание 2
Реализуйте базовую модель логистической регрессии для классификации текстовых сообщений (используемые данные здесь) по признаку спама. Для этого:

1) Приведите весь текст к нижнему регистру;
2) Удалите мусорные символы;
3) Удалите стоп-слова;
4) Привидите все слова к нормальной форме;
5) Преобразуйте все сообщения в вектора TF-IDF. Вам поможет следующий код:

from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=cname)
Можете поэкспериментировать с параметрами TfidfVectorizer;
6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите random_state=42. Используйте train_test_split;
7) Постройте модель логистической регрессии, укажите random_state=42, оцените ее точность на тестовых данных;
8) Опишите результаты при помощи confusion_matrix;
9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).

In [10]:
data = pd.read_csv('spam.txt')

In [11]:
data.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [12]:
import re
import nltk
nltk.download('stopwords')
nltk.download('wordnet')

from nltk.corpus import stopwords
stopwords_set = set(stopwords.words('english'))

from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Vera_Romantsova\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Vera_Romantsova\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [13]:
def normalize_word(string_):
    string_ = string_.lower() #Приводим к нижнему регистру
    string_ = re.sub('[\W_]+',' ', string_) # уберем все служебные символы
    string_ = string_.split() # разделим все на слова
    string_ = [word for word in string_ if word not in stopwords_set] #убираем стоп-слова
    string_ = [wordnet_lemmatizer.lemmatize(word) for word in string_] #приведем слова к леммам
    string_ = ' '.join(map(str, string_)) #преобразуем в строку
    return string_
data['Message'] = data.Message.apply(normalize_word)

In [14]:
data.head()

Category                                            Message
0      ham  go jurong point crazy available bugis n great ...
1      ham                            ok lar joking wif u oni
2     spam  free entry 2 wkly comp win fa cup final tkts 2...
3      ham                u dun say early hor u c already say
4      ham                nah think go usf life around though

In [15]:
# Преобразуем все сообщения в вектора TF-IDF
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(data.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=None)
tfidf_matrix.head()

0     1     2     3     4     5     6     7     8     9     ...  8038  \
0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
4   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

   8039  8040  8041  8042  8043  8044  8045  8046  8047  
0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 8048 columns]

In [16]:
# Разделим данные на тестовые и тренировочные 30\70 
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(tfidf_matrix, data['Category'], test_size=0.30, random_state=42)

In [17]:
# Построим модель логистической регресии, установим random_state=42
lr = LogisticRegression(random_state=42)

In [18]:
lr.fit(X_train, y_train) #Обучаем модель на тренировочных данных

LogisticRegression(random_state=42)

In [19]:
lr.predict(X_test) #Строим прогноз на тестовых данных

array(['ham', 'ham', 'ham', ..., 'ham', 'spam', 'ham'], dtype=object)

In [20]:
# Оценим ее точность на тестовых данных
from sklearn.metrics import accuracy_score
accuracy_score(y_test, lr.predict(X_test))

0.958732057416268

Вы видим, что доля правильных ответов чуть более 0.95

In [21]:
# Опишем результаты при помощи confusion_matrix
from sklearn.metrics import confusion_matrix
cnf_matrix = confusion_matrix(y_test, lr.predict(X_test))
cnf_matrix

array([[1445,    3],
       [  66,  158]], dtype=int64)

Описание матрицы ошибок:
True Positive: мы предсказали 'ham' верно в 1445 случаях.

True Negative: мы предсказали 'spam' в 158 случаях

False Positive: мы предсказали что это 'ham', но на самом деле было 'spam' в 66 случаях. 

False Negative: мы предсказали что это 'spam', но на самом деле было 'ham' в 3 случаях.

In [22]:
# Построим датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).
Original = pd.DataFrame(y_test) #Создаем DF с оригинальными (фактическими) тестовыми данными
Original.columns = ['Category_original']
Original

Category_original
3245               ham
944                ham
1044               ham
2484               ham
812                ham
...                ...
2505               ham
2525              spam
4975               ham
650               spam
4463               ham

[1672 rows x 1 columns]

In [23]:
Prediction = pd.DataFrame(lr.predict(X_test)) #Создаем DF с предсказанными данными
Prediction.columns = ['Category_prediction']
Prediction

Category_prediction
0                    ham
1                    ham
2                    ham
3                    ham
4                    ham
...                  ...
1667                 ham
1668                spam
1669                 ham
1670                spam
1671                 ham

[1672 rows x 1 columns]

In [24]:
Result = Original.join(data, how='left') #Джойним таблицу c оригинальными (фактическими) тестовыми данными к таблице с сообщениями 
Result

Category_original Category  \
3245               ham      ham   
944                ham      ham   
1044               ham      ham   
2484               ham      ham   
812                ham      ham   
...                ...      ...   
2505               ham      ham   
2525              spam     spam   
4975               ham      ham   
650               spam     spam   
4463               ham      ham   

                                                Message  
3245  squeeeeeze christmas hug u lik frndshp den hug...  
944   also sorta blown couple time recently id rathe...  
1044  mmm thats better got roast b better drink 2 go...  
2484                mm kanji dont eat anything heavy ok  
812   ring come guy costume gift future yowifes hint...  
...                                                 ...  
2505  hello boytoy made home constant thought love h...  
2525  free entry 250 weekly comp send word win 80086...  
4975  aiyo u poor thing u dun wan 2 eat u bathe already  
650   1 000 cash 2 000 prize claim call09050000327 c...  
4463  sorry flaked last night shit seriously goin ro...  

[1672 rows x 3 columns]

In [25]:
Result = Result.drop(['Category'], axis=1) #Удаляем столбец с исходной категорией
Result = Result.reset_index(drop = True) #Сбрасываем индекс
Result.head()

Category_original                                            Message
0               ham  squeeeeeze christmas hug u lik frndshp den hug...
1               ham  also sorta blown couple time recently id rathe...
2               ham  mmm thats better got roast b better drink 2 go...
3               ham                mm kanji dont eat anything heavy ok
4               ham  ring come guy costume gift future yowifes hint...

In [26]:
Result1 = pd.concat([Result,Prediction],axis = 1) #Делаем конкатенацию датафреймов
Result1

Category_original                                            Message  \
0                  ham  squeeeeeze christmas hug u lik frndshp den hug...   
1                  ham  also sorta blown couple time recently id rathe...   
2                  ham  mmm thats better got roast b better drink 2 go...   
3                  ham                mm kanji dont eat anything heavy ok   
4                  ham  ring come guy costume gift future yowifes hint...   
...                ...                                                ...   
1667               ham  hello boytoy made home constant thought love h...   
1668              spam  free entry 250 weekly comp send word win 80086...   
1669               ham  aiyo u poor thing u dun wan 2 eat u bathe already   
1670              spam  1 000 cash 2 000 prize claim call09050000327 c...   
1671               ham  sorry flaked last night shit seriously goin ro...   

     Category_prediction  
0                    ham  
1                    ham  
2                    ham  
3                    ham  
4                    ham  
...                  ...  
1667                 ham  
1668                spam  
1669                 ham  
1670                spam  
1671                 ham  

[1672 rows x 3 columns]

In [27]:
# Оставляем только строки отличающиеся предсказанным и исходным значением категории
Result1 = Result1[Result1.Category_original != Result1.Category_prediction]
Result1

Category_original                                            Message  \
17                 ham                                      hey free call   
40                spam  reminder downloaded content already paid goto ...   
47                spam  guess somebody know secretly fancy wanna find ...   
74                spam  oh god found number glad text back xafter msg ...   
84                spam  next amazing xxx picsfree1 video sent enjoy on...   
...                ...                                                ...   
1525              spam  freemsg hi baby wow got new cam moby wanna c h...   
1567              spam    important customer service announcement premier   
1569              spam  themob check newest selection content game ton...   
1576               ham                                          free call   
1586              spam  sm ac blind date 4u rodds1 21 aberdeen united ...   

     Category_prediction  
17                  spam  
40                   ham  
47                   ham  
74                   ham  
84                   ham  
...                  ...  
1525                 ham  
1567                 ham  
1569                 ham  
1576                spam  
1586                 ham  

[69 rows x 3 columns]